src/stdhasher.cr

```diff
@@ -0,0 +1,248 @@
+require "secure_random"
+
+# Hasher usable for `def hash(hasher)` should satisfy protocol:
+#   class MyHasher
+#     def <<(v : Int::Primitive)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(b : Bytes)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(n : Nil)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(v)
+#       v.hash(self)
+#       self
+#     end
+#
+#     # digest returns hashsum for current state without state mutation
+#     # Note: hashsum should implement commutative `+` if MyHasher used
+#     # for `Hasher#hash(hasher)`
+#     def digest
+#     end
+#
+#     def clone_build
+#       with_state_copy do |copy|
+#         yield copy
+#         copy.digest
+#       end
+#     end
+#   end
+
+# StdHasher used as standard hasher in `Object#hash`
+# It have to provide defenense against HashDos, and be reasonably fast.
+# To protect against HashDos, it is seeded with secure random, and have
+# permutation that hard to forge without knowing seed and seeing hash digest.
+#
+# Also it has specialized methods for primitive keys with different seeds.
+struct StdHasher
+  @@seed = StaticArray(UInt32, 7).new{|i| 0_u32}
+  def self.init
+    raise "#{@@seed[0]} #{@@seed[1]}" unless @@seed[0] == 0_u32 && @@seed[1] == 0_u32
+    buf = pointerof(@@seed).as(Pointer(UInt8))
+    SecureRandom.random_bytes(buf.to_slice(sizeof(typeof(@@seed))))
+  end
+
+  private def initialize(@h : Pointer(Impl))
+  end
+
+  def initialize
+    @h = Pointer(Impl).malloc(1, Impl.new(@@seed[0], @@seed[1]))
+  end
+
+  protected def initialize(a, b)
+    @h = Pointer(Impl).malloc(1, Impl.new(a, b))
+  end
+
+  def self.build
+    h = Impl.new(@@seed[0], @@seed[1])
+    s = new(pointerof(h))
+    yield s
+    s.digest
+  end
+
+  def self.hashit(v)
+    build do |s|
+      s << v
+    end
+  end
+
+  protected def self.build(a, b)
+    h = Impl.new(a, b)
+    s = new(pointerof(h))
+    yield s
+    s.digest
+  end
+
+  def clone_build
+    self.class.build(@h.value.a, @h.value.b) do |s|
+      yield s
+    end
+  end
+
+  def clone
+    self.class.new(@h.value.a, @h.value.b)
+  end
+
+  def <<(v : Nil)
+    @h.value.permute_nil(@@seed[2])
+    self
+  end
+
+  def <<(v : Int8|Int16|Int32|UInt8|UInt16|UInt32)
+    @h.value.permute(v.to_u32, @@seed[2])
+    self
+  end
+
+  def <<(v : UInt64)
+    high = (v>>32).to_u32
+    if high != 0
+      self << high
+    end
+    self << v.to_u32
```

Comment:
I have commented `Hashing::Type` but you may introduce it as `alias StdHasher::Type = Int32`.

So please update it, you need these lines for 64bit integers.

```
  private HASH_BITS     =     31 # sizeof(StdHasher::Type) >= 8 ? 61 : 31

    exp = exp >= 0 ? exp % HASH_BITS : HASH_BITS - 1 - ((-1 - exp) % HASH_BITS)
    x = ((x << exp) & HASH_MODULUS) | x >> (HASH_BITS - exp)
```